adapters/richaudience/richaudience.go

```diff
@@ -0,0 +1,349 @@
+
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for _, bid := range seatBid.Bid {
+			b := &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: "banner",
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func setHeaders(raiHeaders *http.Header) {
+	raiHeaders.Set("Content-Type", "application/json;charset=utf-8")
+	raiHeaders.Set("Accept", "application/json")
+	raiHeaders.Add("X-Openrtb-Version", "2.5")
+}
+
+func setImp(request *openrtb2.BidRequest, secure *int8, richaudienceRequest *richaudienceRequest) (raiImp openrtb2.Imp, err error) {
+	//Imp: Id, tagId, Secure, Bidfloor, Bidfloorcur, Banner, test
+	//Banner: Id, W, H, Format
+	//Format: W, H
+
+	for _, imp := range request.Imp {
+		raiExt := parseImpExt(&imp)
+
+		if raiExt != nil {
+			if raiExt.Pid != "" {
+				imp.TagID = raiExt.Pid
+			}
+
+			if raiExt.TestRa {
+				richaudienceRequest.Test = int8(1)
+			}
+
+			if raiExt.BidFloor <= 0 {
+				imp.BidFloor = 0.00001
+			} else {
+				imp.BidFloor = raiExt.BidFloor
+			}
+
+			if raiExt.BidFloorCur == "" {
+				imp.BidFloorCur = "USD"
+			} else {
+				imp.BidFloorCur = raiExt.BidFloorCur
+			}
+		}
+
+		if request.Site != nil && request.Site.Page != "" {
+			pageURL, error := url.Parse(request.Site.Page)
+			if error == nil && pageURL.Scheme == "https" {
+				*secure = int8(1)
+			}
+		}
+		imp.Secure = secure
+
+		if imp.Banner == nil {
+			err = &errortypes.BadInput{
+				Message: "Banner Object not found",
+			}
+			return
+		} else {
+			raiBanner := *imp.Banner
+			if raiBanner.W == nil && raiBanner.H == nil {
+				if len(raiBanner.Format) == 0 {
+					err = &errortypes.BadInput{
+						Message: "Format Object not found",
+					}
+				} else {
+					imp.Banner = &raiBanner
+				}
+			}
+		}
+
+		raiImp = imp
+
+	}
+	return
+}
+
+func setSite(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) {
+	if request.Site.Domain == "" {
+		raiUrl := strings.Split(request.Site.Page, "//")
+		richaudienceRequest.Site.Domain = strings.Split(raiUrl[1], "/")[0]
+	} else {
+		richaudienceRequest.Site.Domain = request.Site.Domain
+	}
+
+	if request.Site.Page != "" {
+		richaudienceRequest.Site.Page = request.Site.Page
+	}
+}
+
+func setDevice(request *openrtb2.BidRequest, richaudienceRequest *richaudienceRequest) (err error) {
+
+	if request.Device.DNT != nil {
+		richaudienceRequest.Device.DNT = *request.Device.DNT
+	} else {
+		richaudienceRequest.Device.DNT = 0
+	}
+
+	if request.Device.Lmt != nil {
+		richaudienceRequest.Device.Lmt = *request.Device.Lmt
+	} else {
+		richaudienceRequest.Device.Lmt = 0
+	}
+
+	if request.Device.UA != "" {
+		richaudienceRequest.Device.UA = request.Device.UA
+	}
+
+	//request.Device.IP = "11.222.33.44"
+
+	if request.Device.IP == "" {
+		err = &errortypes.BadInput{
+			Message: "Not found IP",
```

Comment:
If you make this field required (it's not required in OpenRTB hence request validation will not catch it), please move it to the beginning of the function and make error text more descriptive like "request.Device.IP is required"

Reply:
in next commit i have solved:

I have modifed the text and moving this code to at the beginning function.
For us is important by servers management